src/java.base/share/classes/java/lang/PinnedThreadPrinter.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2020, 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package java.lang;
+
+import java.io.PrintStream;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.WeakHashMap;
+import java.util.stream.Collectors;
+import static java.lang.StackWalker.Option.*;
+
+/**
+ * Helper class to print the virtual thread stack trace when pinned.
+ *
+ * The class maintains a mapping of Class object to the hashes of stack traces that are
+ * pinned by code in that Class. This is used to avoid printing the same stack trace
+ * many times.
+ */
+class PinnedThreadPrinter {
+    static final StackWalker STACK_WALKER;
+    static {
+        var options = Set.of(SHOW_REFLECT_FRAMES, RETAIN_CLASS_REFERENCE);
+        PrivilegedAction<StackWalker> pa = () ->
+            LiveStackFrame.getStackWalker(options, VirtualThread.continuationScope());
+        @SuppressWarnings("removal")
+        var stackWalker = AccessController.doPrivileged(pa);
+        STACK_WALKER = stackWalker;
+    }
+
+    // maps a class to the hashes of stack traces pinned by that code in that class
+    private static final Map<Class<?>, Hashes> classToHashes = new WeakHashMap<>();
```

Comment:
Can you use `ClassValue` in this case?

Reply:
I was wondering that too, but held off commenting since it's not super performance critical given `classToHashes` is synchronized on. However, it does make the code a little clearer.

Reply:
It's not critical and no problem if this is cleaned up as a follow-up.

Reply:
Good idea, this could be improved. As Paul says, it's not performance critical as this is a diagnostic option for printing the stack trace when pinned.